### ISSUE TITLE
Allow specifying changelog headers to look for

### ISSUE THREAD

**Issue by strugee** [hypermodules/gh-release#64]
I'm trying to use this to automate the release of a [large project](https://github.com/pump-io/pump.io). Since it's so large we regularly publish beta releases. This is represented in the changelog as e.g. "5.0.0 beta 0", the corresponding git tag/package.json `version` being `v5.0.0-beta.0`.

Currently this tool thinks the changelog and my package.json are out of sync because of this. It'd be nice if I could specify the header to look for in the changelog (e.g. I could get the correct header with `echo $npm_package_version | sed 's/-beta.\([[:digit:]]\+\)/ beta \1/'`) so I could still keep the logic to check if it's in sync, but not bail out on beta versions.

**Comment by ungoldman** [comment]
Is it too much of a stretch for your project to have the version name match in change log and git? This was originally intended by by design, so that release titles, git tags, and change log headers would all be the same. Also anything you put after the version number is ignored.

**Comment by ungoldman** [comment]
I need to change the changelog parser to accept different heading sizes to be compatible with `standard-version` (see https://github.com/hypermodules/changelog-parser/issues), but loosening the parser to allow non-semver strings to match to version strings seems a little iffy to me.

**Comment by strugee** [comment]
> This was originally intended by design, so that release titles, git tags, and change log headers would all be the same.

Why? Git tags and package.json version numbers are precise things designed for machines, but changelogs are designed to be read by humans. It's not all that big a stretch, but I still think "5.0.0 beta 0" is more readable than "5.0.0-beta.0". Takes less effort to parse and looks more like regular English.

> Also anything you put after the version number is ignored.

Wait, really? This seems super problematic, even from a strict semver perspective - see [this rule in the spec](http://semver.org/#spec-item-9).

> loosening the parser to allow non-semver strings to match to version strings seems a little iffy to me.

I don't feel the same way... I _definitely_ would feel that way if the parser was trying to _automatically_ handle multiple cases, but that's not what I'm suggesting here - I'm suggesting a very intentional override of the default parser behavior.

**Comment by ungoldman** [comment]
**tl;dr:**

- gh-release isn't the place to argue about change log standards
- we stick to the [keep a changelog](http://keepachangelog.com) set of conventions, you should too

---

> Wait, really? This seems super problematic, even from a strict semver perspective - see this rule in the spec.

Anything after the *complete* semver version number is ignored. So if you wanted to you could add a release title that is different from the version afterwards.

```
## 5.0.0-beta.0 - 2017-09-02 - My Cool Title
```

> I don't feel the same way... I definitely would feel that way if the parser was trying to automatically handle multiple cases, but that's not what I'm suggesting here - I'm suggesting a very intentional override of the default parser behavior.

So I guess I wasn't totally clear on what you were asking for -- if I understand correctly now, the request is to create a new option to allow for supplying an alternate regular expression to check if the version in the change log matches the version in package.json, in case the change log author doesn't want to write a semver compliant version string for new versions in the change log.

Though I understand some of your reasoning and don't want to seem unsympathetic, I think this request falls outside of the scope of this project. A big part of the design is based on adhering to an existing change log standard -- http://keepachangelog.com -- which does have a convention around the first part of the release title being a semver compliant version string. A big part of the reasoning for adopting a standard is to avoid [bikeshedding](https://www.freebsd.org/doc/en/books/faq/misc.html#bikeshed-painting), much in the same spirit as the [standard](https://standardjs.com/#i-disagree-with-rule-x-can-you-change-it) js style project.

I would humbly suggest worrying a little less about change log version title styles and just let the standard do the work for you.

Another alternative is to use the `gh-release --title` option with a little of the bash magic you demonstrated above to replace some periods with spaces for the title on the github releases page.

🙏

**Comment by strugee** [comment]
> A big part of the design is based on adhering to an existing change log standard -- http://keepachangelog.com -- which does have a convention around the first part of the release title being a semver compliant version string.

Citation Needed:tm:?

I'm aware of what bikeshedding is. I _do_ stick to keepachangelog and I've never seen anything requiring that section headers have to be semver strings. In fact, keepachangelog [doesn't even imply semver](http://keepachangelog.com/en/1.0.0/#how):

> Mention whether you follow Semantic Versioning. 

I could take this upstream, though?

> Another alternative is to use the `gh-release --title` option with a little of the bash magic you demonstrated above to replace some periods with spaces for the title on the github releases page.

This seems close to what I want, but with a package.json `version` of `5.0.0-beta.1` and the latest entry in the changelog starting with `## 5.0.0 beta 1 - 2017-08-26`, I get:

```
% ./node_modules/.bin/gh-release -n '5.0.0 beta 1'
Error: CHANGELOG.md out of sync with package.json (5.0.0 !== 5.0.0 beta 1)
```

**Comment by ungoldman** [comment]
The `gh-release --name` option sets the title for the github release. If your package.json and change log match `'5.0.0-beta-1' === '5.0.0-beta-1'` it will run the release with the given name. This part maps to the Releases API spec: https://developer.github.com/v3/repos/releases/#create-a-release

I'd really rather not argue about this anymore. If you can come up with a way to modify this tool to suit your needs in an unobtrusive way, a PR is welcome.
